### PR TITLE
[WIP] Expand Asset-buyer2.0 to support Market Sell

### PR DESCRIPTION
**Pull request by dave4506**
## Description

Expanded `SwapQuoter` interface to support both market sell and market buy needs. 

Expanded `SwapQuoteConsumers` to provide outputs/execute market sell `SwapQuotes`

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-buyer/test/utils/swap_quote.ts [hunk 1]
```diff
@@ -56,3 +57,36 @@ export const getFullyFillableSwapQuoteWithNoFees = (
         worstCaseQuoteInfo: quoteInfo,
     };
 };
+
+export const getFullyFillableMarketSellSwapQuoteWithNoFees = (
+    makerAssetData: string,
+    takerAssetData: string,
+    orders: SignedOrder[],
+): MarketSellSwapQuote => {
```
Comment (fragosti) [inline packages/asset-buyer/test/utils/swap_quote.ts]: Same here. Can we DRY this up a bit?

File: packages/order-utils/src/market_utils.ts [hunk 2]
```diff
@@ -9,10 +9,72 @@ import {
...
-    OrdersAndRemainingFillAmount,
+    FindOrdersThatCoverTakerAssetFillAmountOpts,
+    OrdersAndRemainingMakerFillAmount,
+    OrdersAndRemainingTakerFillAmount,
 } from './types';
 
 export const marketUtils = {
+    findOrdersThatCoverTakerAssetFillAmount<T extends Order>(
+        orders: T[],
+        takerAssetFillAmount: BigNumber,
```
Comment (fragosti) [inline packages/order-utils/src/market_utils.ts]: Could we DRY this up as well?

File: packages/order-utils/src/types.ts [hunk 3]
```diff
@@ -37,6 +37,18 @@ export interface FindOrdersThatCoverMakerAssetFillAmountOpts {
...
 }
 
+/**
+ * remainingFillableMakerAssetAmount: An array of BigNumbers corresponding to the `orders` parameter.
+ * You can use `OrderStateUtils` `@0x/order-utils` to perform blockchain lookups for these values.
+ * Defaults to `makerAssetAmount` values from the orders param.
+ * slippageBufferAmount: An additional amount of makerAsset to be covered by the result in case of trade collisions or partial fills.
+ * Defaults to 0
+ */
+export interface FindOrdersThatCoverTakerAssetFillAmountOpts {
```
Comment (fragosti) [inline packages/order-utils/src/types.ts]: nit: a bit odd that this is named like a function with `Find`

File: yarn.lock [hunk 4]
```diff
@@ -478,17 +478,17 @@
   dependencies:
     "@0x/base-contract" "^4.0.3"
 
-"@0x/asset-buyer@6.1.5", "@0x/asset-buyer@^6.1.5":
-  version "6.1.5"
-  resolved "https://registry.yarnpkg.com/@0x/asset-buyer/-/asset-buyer-6.1.5.tgz#bf2056ccd9e749034d993c4de259d2f1b572421e"
-  integrity sha512-8e2OTSEacyWnWiYtchKHeQ3mYq912dd8OL46Wekt7DMNYx85R1ExTHaTgoKEQAe4EDsETW41yBPT6DLXwMI/Tw==
+"@0x/asset-buyer@6.1.4":
+  version "6.1.4"
+  resolved "https://registry.yarnpkg.com/@0x/asset-buyer/-/asset-buyer-6.1.4.tgz#ea863b860fbae6f633846bdcf23cacbb345aefd1"
```
Comment (fragosti) [inline yarn.lock]: Why the downgrade here?

Comment (dave4506) [reply yarn.lock]: This probably had to do with getting CI to pass with a forced downgrade.

File: packages/asset-buyer/src/types.ts [hunk 5]
```diff
@@ -183,9 +201,10 @@ export interface SwapQuoteWithAffiliateFee extends SwapQuote {
  * totalEthAmount: The total amount of eth required to complete the buy (filling orders, feeOrders, and paying affiliate fee).
  */
 export interface SwapQuoteInfo {
-    takerTokenAmount: BigNumber;
     feeTakerTokenAmount: BigNumber;
     totalTakerTokenAmount: BigNumber;
+    takerTokenAmount: BigNumber;
```
[the remark this answers is not in the document]
Comment (dave4506) [reply packages/asset-buyer/src/types.ts]: yup, total is fee + taker

File: packages/order-utils/src/types.ts (hunk 3 above)
Comment (dave4506) [reply packages/order-utils/src/types.ts]: Yeah, not a fan of it but the convention for the opts for the other functions fall under this pattern, open to changing this and the convention.

File: packages/asset-buyer/src/swap_quoter.ts [hunk 6]
```diff
@@ -130,20 +186,20 @@ export class SwapQuoter {
...
      */
-    public async getSwapQuoteAsync(
+    public async getMarketBuySwapQuoteAsync(
         makerAssetData: string,
         takerAssetData: string,
-        makerAssetSwapAmount: BigNumber,
+        makerAssetBuyAmount: BigNumber,
         options: Partial<SwapQuoteRequestOpts> = {},
-    ): Promise<SwapQuote> {
+    ): Promise<MarketBuySwapQuote> {
```
Comment (fragosti) [inline packages/asset-buyer/src/swap_quoter.ts]: Can we de-dupe `getMarketBuySwapQuoteAsync` and `getMarketSellSwapQuoteAsync` as well?

File: packages/order-utils/src/types.ts [hunk 7]
```diff
@@ -37,6 +37,20 @@ export interface FindOrdersThatCoverMakerAssetFillAmountOpts {
...
+    remainingFillableTakerAssetAmounts?: BigNumber[];
+    slippageBufferAmount?: BigNumber;
+}
+
+export type MarketOperation = 'marketSell' | 'marketBuy';
```
Comment (fragosti) [inline packages/order-utils/src/types.ts]: Is this intended to be duplicate?

Comment (fragosti) [reply packages/order-utils/src/types.ts]: If it is, I'd say to move it to `@0x/types`

File: packages/asset-buyer/src/types.ts [hunk 8]
```diff
@@ -163,29 +177,44 @@ export interface ForwarderSwapQuoteExecutionOpts extends ForwarderSwapQuoteGetOu
...
     worstCaseQuoteInfo: SwapQuoteInfo;
 }
 
-export interface SwapQuoteWithAffiliateFee extends SwapQuote {
+export interface MarketSellSwapQuote extends SwapQuoteBase {
+    takerAssetFillAmount: BigNumber;
+    type: 'marketSell';
+}
+
+export interface MarketBuySwapQuote extends SwapQuoteBase {
```
Comment (fragosti) [inline packages/asset-buyer/src/types.ts]: 👍

File: packages/asset-buyer/src/utils/swap_quote_calculator.ts [hunk 9]
```diff
@@ -4,131 +4,239 @@ import * as _ from 'lodash';
...
+        return calculateSwapQuote(
+            ordersAndFillableAmounts,
+            feeOrdersAndFillableAmounts,
+            takerAssetFillAmount,
+            slippagePercentage,
+            isMakerAssetZrxToken,
+            'marketSell',
+        ) as MarketSellSwapQuote;
+    },
+    calculateMarketBuySwapQuote(
```
Comment (fragosti) [inline packages/asset-buyer/src/utils/swap_quote_calculator.ts]: 🙏

**Pull request merged**